### ISSUE TITLE
[TMVA][SOFIE] Status of the support of ONNX operators

### ISSUE THREAD

**Issue by axmat** [root-project/root#10360]
This issue should track the status of the support of ONNX operators in TMVA/SOFIE.

Supported :

- [x] Add
- [x] BatchNormalization
- [x] Conv
- [x] Gemm
- [x] GRU
- [x] LSTM
- [x] Pool
- [x] Relu
- [x] Reshape
- [x] RNN
- [x] Selu
- [x] Sigmoid
- [x] Slice
- [x] Transpose

Wok in progress :

- [ ] Softmax, #10315 
- [ ] Identity, #10315 
- [ ] Concat, #10315 and #9311 
- [ ] InstanceNormalization, #8885

Missing :

- [ ] Abs
- [ ] Acos
- [ ] Acosh
- [ ] And
- [ ] ArgMax
- [ ] ArgMin
- [ ] Asin
- [ ] Asinh
- [ ] Atan
- [ ] Atanh
- [ ] AveragePool
- [ ] BitShift
- [ ] Cast
- [ ] Ceil
- [ ] Clip
- [ ] Compress
- [ ] ConcatFromSequence
- [ ] Constant
- [ ] ConstantOfShape
- [ ] ConvInteger
- [ ] ConvTranspose
- [ ] Cos
- [ ] Cosh
- [ ] CumSum
- [ ] DepthToSpace
- [ ] DequantizeLinear
- [ ] Det
- [ ] Div
- [ ] Dropout
- [ ] Einsum
- [ ] Elu
- [ ] Equal
- [ ] Erf
- [ ] Exp
- [ ] Expand
- [ ] EyeLike
- [ ] Flatten
- [ ] Floor
- [ ] Gather
- [ ] GatherElements
- [ ] GatherND
- [ ] GlobalAveragePool
- [ ] GlobalLpPool
- [ ] GlobalMaxPool
- [ ] Greater
- [ ] GridSample
- [ ] HardSigmoid
- [ ] Hardmax
- [ ] Identity
- [ ] If
- [ ] IsInf
- [ ] IsNaN
- [ ] LRN
- [ ] LeakyRelu
- [ ] Less
- [ ] Log
- [ ] Loop
- [ ] LpNormalization
- [ ] LpPool
- [ ] MatMul
- [ ] MatMulInteger
- [ ] Max
- [ ] MaxPool
- [ ] MaxRoiPool
- [ ] MaxUnpool
- [ ] Mean
- [ ] Min
- [ ] Mod
- [ ] Mul
- [ ] Multinomial
- [ ] Neg
- [ ] NonMaxSuppression
- [ ] NonZero
- [ ] OneHot
- [ ] Optional
- [ ] OptionalGetElement
- [ ] OptionalHasElement
- [ ] Or
- [ ] PRelu
- [ ] Pad
- [ ] Pow
- [ ] QLinearConv
- [ ] QLinearMatMul
- [ ] QuantizeLinear
- [ ] RandomNormal
- [ ] RandomNormalLike
- [ ] RandomUniform
- [ ] RandomUniformLike
- [ ] Reciprocal
- [ ] ReduceL1
- [ ] ReduceL2
- [ ] ReduceLogSum
- [ ] ReduceLogSumExp
- [ ] ReduceMax
- [ ] ReduceMean
- [ ] ReduceMin
- [ ] ReduceSum
- [ ] ReduceSumSquare
- [ ] Resize
- [ ] ReverseSequence
- [ ] RoiAlign
- [ ] Round
- [ ] Scan
- [ ] ScatterElements
- [ ] ScatterND
- [ ] SequenceAt
- [ ] SequenceConstruct
- [ ] SequenceEmpty
- [ ] SequenceErase
- [ ] SequenceInsert
- [ ] SequenceLength
- [ ] Shape
- [ ] Shrink
- [ ] Sign
- [ ] Sin
- [ ] Sinh
- [ ] Size
- [ ] Softplus
- [ ] Softsign
- [ ] SpaceToDepth
- [ ] Split
- [ ] SplitToSequence
- [ ] Sqrt
- [ ] StringNormalizer
- [ ] Sub
- [ ] Sum
- [ ] Tan
- [ ] Tanh
- [ ] TfldfVectorizer
- [ ] ThresholdedRelu
- [ ] Tile
- [ ] TopK
- [ ] Trilu
- [ ] Unique
- [ ] Unsqueeze
- [ ] Where
- [ ] Xor

**Comment by youlxs** [comment]
Do we have the plan to support "MatMul" ?

**Comment by lmoneta** [comment]
MatMul is now supported in both master and 6.30, see

https://github.com/root-project/root/pull/12894

**Comment by youlxs** [comment]
Do we have a plan to support "Split" ?

**Comment by lmoneta** [comment]
We can plan to support it if needed. It might take a little bit of time, since we need to support in SOFIE sequence of tensors which are the output of Split. For which type of model do you need Split?

**Comment by youlxs** [comment]
> MatMul is now supported in both master and 6.30, see
> 
> #12894
> 
> This above list needs to be updated with the new supported operator

thanks

**Comment by youlxs** [comment]
> We can plan to support it if needed. It might take a little bit of time, since we need to support in SOFIE sequence of tensors which are the output of Split. For which type of model do you need Split?

I want to convert my model to the raw code to improve the prediction speed. As I saw from your slide the raw code generated through tmva has better performance than the original onnx runtime.

**Comment by Shlok-Saxena** [comment]
Hi, I saw this issue about documenting ONNX operator support in SOFIE.  I’ve been working with TMVA SOFIE parsers recently and added documentation that reflects the operators currently registered in the codebase.

I’ve opened a PR with these changes: #20811

**Comment by Shlok-Saxena** [comment]
Hi, I worked on documenting the supported and missing ONNX operators in SOFIE parsers and opened PR #20811 for it.

Initially, the operator list was added as a `README.md` in `root/tmva/sofie_parsers`. Since I’ve now opened a separate PR #20825 to introduce the main `root/tmva/sofie_parsers/README.md`, I’m going to move the operator list into a dedicated document `ONNX_OPERATOR_STATUS.md` under `root/tmva/sofie` directory and link it from the existing `root/tmva/sofie/README.md` — instead of keeping it in the parsers `README`.  

This avoids overlapping changes and keeps the `README` focused.

Feedback on the preferred location and format for the operator list is welcome!

**Comment by lmoneta** [comment]
The list of ONNX supported operators is in the README fiel of the tmva/sofie director. 
Close this issue since it has not been updated since a long time
